### ISSUE TITLE
Missing TraceInfo for requests with tracing, save response, and download callback enabled

### ISSUE THREAD

**Issue by cheeseprocedure** [imroc/req#200]
# Summary

When a request has tracing enabled, is configured to save output (e.g. by calling `SetOutput()`), and configured with a download callback, HTTP trace hooks in the request context appear to be lost, resulting in mostly zero values for the request's `TraceInfo` fields.

# Details

In [`client.go` v3.19.1](https://github.com/imroc/req/blob/v3.19.1/client.go#L1269) (the version I recently updated from), the request context is initialized and passed to the HTTP request like so, and I appear to get valid `TraceInfo` values when the request completes:

```go
if r.trace == nil && r.client.trace {
	r.trace = &clientTrace{}
}
if r.trace != nil {
	r.ctx = r.trace.createContext(r.Context())
}

// ...
ctx := r.ctx

if r.isSaveResponse && r.downloadCallback != nil {
	// ...
	if ctx == nil {
		ctx = context.Background()
	}
	ctx = context.WithValue(ctx, wrapResponseBodyKey, wrap)
}
if ctx != nil {
	req = req.WithContext(ctx)
}
```

...but in [v3.26.6](https://github.com/imroc/req/blob/v3.26.6/client.go#L1299), `ctx` is conditionally overwritten using _the `*req.Request`'s original context_ instead of the existing `ctx`, which at this point may have trace hooks registered:

```go
if r.trace == nil && r.client.trace {
	r.trace = &clientTrace{}
}
ctx := r.ctx
if r.trace != nil {
	ctx = r.trace.createContext(r.Context())
}

// ...

if r.isSaveResponse && r.downloadCallback != nil {
	// ...
	ctx = context.WithValue(r.Context(), wrapResponseBodyKey, wrap)
}

if ctx != nil {
	req = req.WithContext(ctx)
}
```

With this version of the package, I see mostly zero values in the request's `TraceInfo`, and logging the output of `httptrace.ContextClientTrace(ctx)` reports `(*httptrace.ClientTrace)(nil)`, confirming trace hooks are not present.

It appears this change was introduced [while addressing an issue identified by Go's data race detector](https://github.com/imroc/req/commit/5a3fb0994710ea09f23b025031b22948a7dfcb90) (#159). 

# Potential resolution

I've verified that changing the conditional line to `ctx = context.WithValue(ctx, wrapResponseBodyKey, wrap)` resolves the issue locally, and I cannot reproduce the data race from #159 after this patch, but I'm not certain it's a complete fix.

**Thank you** for your work on this project!

# Environment

**Go version:** `go version go1.19.3 darwin/amd64`

**Comment by imroc** [comment]
Thanks for the feedback, this is a very hidden bug, `ctx` might be nil at that line, so use `r.Context()` to get `ctx`, but unfortunately, `ctx` might be overwritten when trace is enabled and download callback is set.

This is fixed in [this commit](https://github.com/imroc/req/commit/9d5943a7cbf125e358cf57bc56b47bb4a8596d1e), you can upgrade to the latest version.

**Comment by cheeseprocedure** [comment]
Thanks so much! I've confirmed the issue is resolved for us in `v3.26.7`.
